File: pkg/model/rule.go
package model

import (
	"go.mongodb.org/mongo-driver/v2/bson"
)

// RuleType 规则类型
//
//	@Description	规则类型，表示规则是白名单还是黑名单
type RuleType string

const (
	WhitelistRule RuleType = "whitelist" // 白名单规则
	BlacklistRule RuleType = "blacklist" // 黑名单规则
)

// RuleStatus 规则状态
//
//	@Description	规则状态，表示规则是启用还是禁用
type RuleStatus string

const (
	RuleEnabled  RuleStatus = "enabled"  // 规则已启用
	RuleDisabled RuleStatus = "disabled" // 规则已禁用
)

// MicroRule 表示WAF微规则信息
// @Description WAF微规则信息，包含规则名称、类型、状态、优先级和条件
type MicroRule struct {
	ID       bson.ObjectID `bson:"_id,omitempty" json:"id,omitempty" example:"60d21b4367d0d8992e89e964"` // 规则唯一标识符
	Name     string        `json:"name" bson:"name" example:"SQL注入防护规则"`                                 // 规则名称
	Type     RuleType      `json:"type" bson:"type" example:"blacklist"`                                 // 规则类型
	Status   RuleStatus    `json:"status" bson:"status" example:"enabled"`                               // 规则状态
	Priority int           `json:"priority" bson:"priority" example:"100"`                               // 优先级字段，数字越大优先级越高
	// @Schema(type=object, example={"type":"composite","operator":"AND","conditions":[{"type":"simple","target":"source_ip","match_type":"in_ipgroup","match_value":"blocked_ips"},{"type":"simple","target":"path","match_type":"regex","match_value":"^/admin/.*$"}]})
	Condition bson.Raw `json:"condition" bson:"condition" swaggertype:"object"`
}

func (r *MicroRule) GetCollectionName() string {
	return "micro_rule"
}

File: web/src/types/rule.ts
// 匹配目标类型
export type TargetType = 'source_ip' | 'url' | 'path'

// 匹配方式类型
export type MatchType =
    // IP匹配方式
    | 'equal'
    | 'not_equal'
    | 'fuzzy'
    | 'in_cidr'
    | 'not_in_cidr'
    | 'in_ipgroup'
    | 'not_in_ipgroup'
    // URL和Path匹配方式
    | 'include'
    | 'contains'
    | 'not_contains'
    | 'prefix_keyword'
    | 'regex'

// 逻辑操作符
export type LogicalOperator = 'AND' | 'OR'

// 条件类型
export type ConditionType = 'simple' | 'composite'

// 规则类型
export type RuleType = 'whitelist' | 'blacklist'

// 规则状态
export type RuleStatus = 'enabled' | 'disabled'

// 简单条件
export interface SimpleCondition {
    type: 'simple'
    target: TargetType
    match_type: MatchType
    match_value: string
}

// 复合条件
export interface CompositeCondition {
    type: 'composite'
    operator: LogicalOperator
    conditions: (SimpleCondition | CompositeCondition)[]
}

// 条件类型(联合类型)
export type Condition = SimpleCondition | CompositeCondition

// 微规则模型
export interface MicroRule {
    id: string
    name: string
    type: RuleType
    status: RuleStatus
    priority: number
    condition: Condition
    createdAt?: string
    updatedAt?: string
}

// 创建规则请求
export interface MicroRuleCreateRequest {
    name: string
    type: RuleType
    status: RuleStatus
    priority: number
    condition: Condition
}

// 更新规则请求
export interface MicroRuleUpdateRequest {
    name?: string
    type?: RuleType
    status?: RuleStatus
    priority?: number
    condition?: Condition
}

// 规则列表响应
export interface MicroRuleListResponse {
    total: number
    items: MicroRule[]
}

// 目标类型与匹配方式的映射关系
export const TARGET_MATCH_TYPES: Record<TargetType, MatchType[]> = {
    'source_ip': ['equal', 'not_equal', 'fuzzy', 'in_cidr', 'not_in_cidr', 'in_ipgroup', 'not_in_ipgroup'],
    'url': ['equal', 'not_equal', 'contains', 'not_contains', 'prefix_keyword', 'regex'],
    'path': ['equal', 'not_equal', 'contains', 'not_contains', 'prefix_keyword', 'regex']
}